lib/neovim-plugin.nix

```diff
@@ -94,25 +99,90 @@ with lib;
             example = settingsExample;
           };
         }
+        // optionalAttrs allowLazyLoad {
+          lazyLoad = helpers.mkLazyLoadOption {
+            inherit originalName;
+            lazyLoadDefaults =
+              (optionalAttrs (isColorscheme && colorscheme != null) { inherit colorscheme; }) // lazyLoadDefaults;
+          };
+        }
         // extraOptions;
 
       config =
         let
-          cfg = config.${namespace}.${name};
           extraConfigNamespace = if isColorscheme then "extraConfigLuaPre" else "extraConfigLua";
+          lazyLoaded = cfg.lazyLoad.enable or false;
+
+          # lz-n lazyLoad backend
+          # Transform plugin into attrset and set optional to true
+          # See `tests/test-sources/plugins/pluginmanagers/lz-n.nix`
+          optionalPlugin =
+            x:
+            if isColorscheme then x else ((if x ? plugin then x else { plugin = x; }) // { optional = true; });
+          mkFn = str: if str != "" then "function()\n" + str + "end" else null;
         in
         mkIf cfg.enable (mkMerge [
-          {
-            extraPlugins = (optional installPackage cfg.package) ++ extraPlugins;
-            inherit extraPackages;
-          }
-          (optionalAttrs callSetup {
-            ${extraConfigNamespace} = ''
-              require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
-            '';
-          })
+          # Always set
           (optionalAttrs (isColorscheme && (colorscheme != null)) { colorscheme = mkDefault colorscheme; })
-          (extraConfig cfg)
+
+          # Normal loading
+          (mkIf (!lazyLoaded) (mkMerge [
+            (extraConfig cfg)
+            {
+              extraPlugins = (optional installPackage cfg.package) ++ extraPlugins;
+              inherit extraPackages;
+            }
+            (optionalAttrs callSetup {
+              ${extraConfigNamespace} = ''
+                require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
+              '';
+            })
+          ]))
+
+          # Lazy loading with lz-n
+          (mkIf (lazyLoaded && config.plugins.lz-n.enable) (mkMerge [
+            (lib.removeAttrs (extraConfig cfg) [
+              "extraConfigLua"
+              "extraConfigLuaPre"
+            ])
```

Comment:
This isn't a robust solution, we really need something like #1876 so that we can assign plugin-config to an option and then decide here how to _use_ that option.

The problem is we're assuming `extraConfig` returns a "plain" attrset, instead of something like a `mkIf`, `mkOverride`, `mkMerge`, etc.

It's also kinda "pulling the rug out" from the module definition, which is expecting a config definition to actually add a definition to the respective option.

If we have an intermediate, plugin-specific option, this is more transparent and also exposes better hooks to end-users.